Add dissect.disc

MaxGroot opened the pull request:
This PR adds support for filesystems typically encountered on optical discs. It currently supports ISO 9660 and its extensions 'Rock Ridge' and 'Joliet'. It currently does not support UDF and Apple extensions to ISO9660.

For example, the following usage:

```python
# First insert the best game ever in your CD drive
from dissect.disc import DISC

with open("/dev/sr0", "rb") as best_game_ever:
    disc = DISC(best_game_ever)
    print(disc.volume_name)
    best_day_ever = disc.primary_volume.creation_date
    print(best_day_ever)
    for best_file_ever in disc.get_entry("/").scandir():
        print(best_file_ever.name)
```
Would yield:
```
LEGOISLAND
<dec_datetime year=b'1997', month=b'08', day=b'11', hour=b'22', minute=b'24', second=b'32', centiseconds=b'00', offset=0x30>
.
..
ATTSETUP.EXE
AUTORUN.INF
CONNECT.EX_
DATA
INSTALL.EXE
LEGO
```

MaxGroot committed:
Add dissect.disc

Schamper commented on dissect/disc/c_iso_9660.py:
Can you replace all tabs with spaces and fix the structure field alignments?

Schamper commented on dissect/disc/c_iso_9660.py:
```diff
@@ -0,0 +1,135 @@
+
+ISO_DEF = """
+
+/* volume descriptor types */
```
```suggestion
iso_def = """
/* Volume descriptor types */
```
Not bad per se, but more consistent with other projects.

Schamper commented on dissect/disc/c_iso_9660.py:
```diff
@@ -0,0 +1,135 @@
+"""  # noqa
+
+c_iso = cstruct()
+c_iso.load(ISO_DEF)
```
```suggestion
c_iso.load(iso_def)
```

Schamper commented on dissect/disc/c_iso_9660.py:
```diff
@@ -0,0 +1,135 @@
+    uint16_t		parent_dir_no;
+    char			name[directory_id_len];
+}
+"""  # noqa
```
```suggestion
"""  # noqa: E501
```

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+
+        UDF:
+            - https://docplayer.net/237090850-Ecma-tr-universal-disk-format-udf-specification-part-3-revision-2-50-1st-edition-december-reference-number-ecma-123-2009.html
+
```
Did you check if this renders as expected with `tox -e docs-build`?

MaxGroot replied:
While checking I found that the `tests/docs` directory missed some files. Should be all good now.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+logger = logging.getLogger(__name__)
+
+
+class ISOFormats(Enum):
```
```suggestion
class ISOFormat(Enum):
```

MaxGroot replied:
Renamed to `DiscFormat`

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+]
+
+
+class DISC:
```
The implementation is currently mostly ISO instead of something "generic" that could support both ISO and e.g. UDF. Would it not make sense to name this (and the file) ISO? Then once you add support for UDF, you can look into making something generic.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+        """Load volume descriptors from the ISO file."""
+        self.fh.seek(c_iso.SYSTEM_AREA_SIZE)
+
+        while True:
```
If I accidentally throw in something that's not an ISO file, will this read through the entire file? Is there no (sensible) upper limit on how far into the file to read?

MaxGroot replied:
Addressed by checking for a magic value while iterating.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+    NotASymlinkError,
+)
+
+logger = logging.getLogger(__name__)
```
Can you do this similar to e.g. https://github.com/fox-it/dissect.hypervisor/blob/main/dissect/hypervisor/disk/vmdk.py#L21-L22

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+            )
+
+    def _check_rockridge(self) -> None:
+        """Check whether this disc is rockridge compatible."""
```
```suggestion
        """Check whether this disc is Rockridge compatible."""
```

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+            return RockRidgeDirectoryRecord(self, record)
+        return DirectoryRecord(self, record)
+
+    def get_entry(self, path: str, use_path_table: bool = False) -> DirectoryRecord:
```
```suggestion
    def get(self, path: str, use_path_table: bool = False) -> DirectoryRecord:
```
More consistent with other Dissect libraries.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+
+    @property
+    def ctime(self) -> datetime:
+        """Plain ISO9660 have only one timestamp associated with them."""
```
Can you make the `"""` docstring a description of what `ctime` does and the text you currently have there a `#` comment? Also goes for the other functions.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+        raise NotImplementedError("Extended attribute record is available but not supported")
+
+    @property
+    def user_id(self) -> int:
```
```suggestion
    def uid(self) -> int:
```

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+        raise NotImplementedError("Extended attribute record is available but not supported")
+
+    @property
+    def group_id(self) -> int:
```
```suggestion
    def gid(self) -> int:
```

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+        raise NotImplementedError("Extended attribute record is available but not supported")
+
+    @property
+    def links(self) -> int:
```
```suggestion
    def nlinks(self) -> int:
```
?

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+        if not self.is_dir:
+            self.name, _, _ = self.name.partition(";")
+
+    def scandir(self) -> Iterator[DirectoryRecord]:
```
Can you add a `listdir` that returns a `dict[str, DirectoryRecord]` of `name: record`? E.g. https://github.com/fox-it/dissect.jffs/blob/main/dissect/jffs/jffs2.py#L288-L289

Schamper replied:
Also in other libraries we call this `iterdir`. Not sure why, but might be nice to stay consistent with that?

Schamper commented on pyproject.toml:
```diff
@@ -0,0 +1,52 @@
+requires-python = "~=3.9"
+license.text = "Affero General Public License v3"
+authors = [
+  {name = "Dissect Team", email = "dissect@fox-it.com"}
```
Feel free to add yourself if you want.

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+
+    """  # noqa: E501
+
+    def __init__(self, fh: BinaryIO, preference: Optional[ISOFormats] = None) -> None:
```
Can you add a `Args:` docstring?

Schamper replied:
```suggestion
    def __init__(self, fh: BinaryIO, preference: Optional[ISOFormats] = None):
```

Schamper commented on dissect/disc/disc.py:
```diff
@@ -0,0 +1,756 @@
+from datetime import datetime, timedelta, timezone
+from enum import Enum
+from io import BytesIO
+from typing import BinaryIO, Iterator, Optional, Union
```
```suggestion
from typing import BinaryIO, Iterator, Optional
```
Can you change all `Union` to use the new `|` syntax?

MaxGroot replied:
In addition to this, I also replaced all `Optional` with a ` | None`

Schamper reviewed (changes requested):
Initial few comments, let's first discuss these before I go more in depth and you have to go grab the.. well you know. <sup><sup><sup>miss you buddy</sup></sup></sup>

Two overall comments to think about:
- `dissect.disc` makes sense to support different disc formats in the long run. Currently you support ISO, but call it DISC. Maybe just put all the ISO stuff in `dissect.disc.iso` to make the separation clear.
- Maybe think of a way to separate the extensions into different files. Circular dependency will be an issue though, so you may have to introduce a base file for the base directory record implementation.

Schamper commented:
Can you also include these two changes? 
https://github.com/fox-it/dissect/commit/72b24953d9e24bf81fe5c547a7b3d84f2c845709
https://github.com/fox-it/dissect/commit/7228e49761f5563cc5462c717aa71f728ab039ad

MaxGroot committed:
Implement review suggestions

MaxGroot committed:
Add docs files

MaxGroot committed:
Check every volume descriptor for ISO9660-magic

MaxGroot committed:
Refactor disc.py into seperate files

MaxGroot commented:
I've addressed your initial few comments in the first three commits, and the broader remarks in the last one. 

Locally, I've started working on UDF, also to further understand what might be a nice structuring for this PR that keeps future extension in mind. I ended up implementing some 'base' logic in `dissect.disc.base`, which can then be extended by different 'disc formats', such as RockRidge, Joliet, 'plain' ISO9660 or UDF later down the line. The `DISC` class is then only  responsible for selecting the right format (with user preference in mind), and providing access to the underlying available disc formats and their APIs.  The abstract classes `DiscBase` and `DiscBaseEntry` implement an almost identical API as `dissect.target`.  This also removes the complicated 'stateful' initialization that the `DISC` class had.

I do have some reservations about the new module structure, but am curious to hear what you think. Perhaps it's acceptable enough that we go more in depth, no holding back. <sup>like we always used to do</sup>

MaxGroot committed:
Compatibility with cstruct v4

MaxGroot committed:
Abstraction layer improvements & bugfixes

- Throw the project-specific exceptions rather than the base python ones
- Handle non-ISO9660 input files correctly
- Move file finding logic to base.py
- Bugfix current/parent directory for Joliet
- Bugfix filesize
- New tests to test the bugfixes

MaxGroot committed:
Refactor out the SUSP class for Rockridge

MaxGroot committed:
Add UDF Support

MaxGroot commented:
Hi @Schamper,

I know you're upset with me for missing your birthday, but I hope this small present makes up for it. I know your favourite activity is to do code review, so I extended this PR with UDF support so there are more lines in this PR for you to play around with. To make it even more challenging, I only briefly looked at consistent spacing and that stuff, because I know how excited you get clicking that sweet 'add suggestion' button.

I think this more than makes up for all those "How could you miss my birthday", "It was really important to me" and "Why are you so obsessed with an optical medium of the 90s" comments/jokes you were making. You're welcome!

MaxGroot committed:
Copycat dissect.cramfs

MaxGroot committed:
Switch to ruff

MaxGroot committed:
Add stubs

MaxGroot commented:
Hi @Schamper,

Strangely enough, I didn't receive an invite for your birthday this year? It almost had me believing that you may not be interested in optical disks anymore. I even wondered if you still review filesystem PRs these days, but then saw https://github.com/fox-it/dissect.cramfs/pull/1 .

I understand. That PR uses Ruff, has stubs for the cstruct definitions and the easter eggs in the PRs are funnier than I could come up with for this one. Coincidentally, I couldn't help noticing that @Horofic was invited for your birthday. 

It's okay, cramfs is the 'new kid on the block', having been released in 2001 compared to UDF's 1995. It seems like only yesterday when I was cruising in my 1995 Renault Twingo listening to Springsteen's just-released 'Greatest Hits' album.

But those days are over. The time of the Renault Clio and Britney Spears has come. 

I've added a few commits to try to keep up with the cool kids, but understand if you can't make time anymore for an old man like me. If you ever want to, as you kids say, 'hang out', feel free to hit me up on my pager.

Kind regards,

Max Groot

Schamper committed:
Suggested changes

Schamper commented:
Hi @MaxGroot,

Sorry, your invitation must've gotten mixed up in the pager network. Not like I sent out 847 other invites. I still care deeply for optical disks, in fact, I was deeply saddened when I learned that they no longer give you a CD when you develop analog pictures. Like, hellooo, who audited that cloud environment you keep my naughty pictures on? Pinky promise like Snapchat to delete them after a week. I'll take them on a soon-to-be coaster, thankyouverymuch.

I've also been a bit sad we haven't seem to be able to keep the romance going like the good old days. We said it wasn't going to be a big deal, that it would suck not working together anymore, but that we'd still see each other all the time. Now every time we speak to each other you seem more interested in the girls on TV. And we don't even get blueberries. I miss your skinny white butt.

Anyway, I've gone through your code, I really like it. I gave it the old `schamperlint` and it seemed to have skipped `udf.py` for now, maybe the 2026 release version works on that. Let me know what you think of these changes so far. For what it's worth, I like reviewing your stuff a lot more than whatever "cramfs" is, and I formally apologize for not taking it more seriously.

Looking forward to continue working on this with you and finally merging all of this.

Kind regards,
Erik

Schamper committed:
More suggestions

Schamper commented:
```text
❯ uv tool upgrade schamperlint
Updated schamperlint v2015.12 -> v2016.01
 - schamperlint==2015.12
 + schamperlint==2016.01
Installed 1 executable: schamperlint
```

I upgraded and ran it again, it made some more suggested changes, though sometimes I'm not sure if it's for better or worse. Let me know what you think!

I've also gone ahead and enabled LFS, so you can dump those big loads in here now too.

Schamper commented:
After all this time it might seem anti-climactic to merge this just like that. I couldn’t wait any longer for your review, just like I can’t wait any longer for you to embrace me once more.

I was actually still working on adding some more stuff, but there were urgent reasons to merge this right now. You see, there might be a very small chance ~our love story~ this repository could get pressed into an actual, real life ancient media format. There were deadlines involved, and I acted on instinct. You know how well I do under pressure.

I happen to be falling into some free time soon. I hope to finish adding support for more ancient file formats no one asked for, and I hope you will find it in your heart to reply to me. I’m not sure how much longer I can hold out without reading your comments before going to sleep. I yearn for it every night.

Hugs and kisses,
Erik
